File: packages/agent_runtime/src/insightgraph_agent/verifier.py
from __future__ import annotations

import json
import logging

import litellm

logger = logging.getLogger(__name__)

VERIFIER_SYSTEM_PROMPT = """You are a verification agent for InsightGraph. Your job is to verify an analysis by checking:

1. Every key finding has supporting evidence from the retrieved data.
2. No contradictions exist between findings.
3. The confidence score is appropriate given the evidence.
4. Citations are accurate.

Respond in JSON:
{
  "verified": true,
  "issues": [],
  "adjusted_confidence": 0.85,
  "evidence_coverage": 0.9,
  "notes": "optional notes"
}

If issues are found:
{
  "verified": false,
  "issues": ["Finding X has no supporting evidence", "Contradiction between..."],
  "adjusted_confidence": 0.4,
  "evidence_coverage": 0.5,
  "notes": "..."
}
"""


class Verifier:
    """Checks conclusions for evidence support, contradictions, and confidence."""

    def __init__(self, model: str = "gpt-4o-mini", api_key: str = ""):
        self.model = model
        self.api_key = api_key

    async def verify(self, analysis: dict, retrieval_results: list[dict]) -> dict:
        """Verify an analysis against the retrieved evidence."""
        context = json.dumps(
            {"analysis": analysis, "evidence": retrieval_results},
            indent=2,
            default=str,
        )

        kwargs: dict = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": VERIFIER_SYSTEM_PROMPT},
                {"role": "user", "content": f"Verify this analysis:\n{context}"},
            ],
            "response_format": {"type": "json_object"},
            "temperature": 0.0,
        }
        if self.api_key:
            kwargs["api_key"] = self.api_key

        try:
            response = await litellm.acompletion(**kwargs)
            content = response.choices[0].message.content
            return json.loads(content)
        except Exception:
            logger.warning("Verifier failed", exc_info=True)
            return {
                "verified": False,
                "issues": ["Verification failed due to an error"],
                "adjusted_confidence": 0.0,
                "evidence_coverage": 0.0,
                "notes": "Verification unavailable",
            }

File: packages/agent-runtime/src/verifier.ts
import type { LLMClient } from "@insightgraph/core";
import { createLLMClient, chatJSON, safeParseLlmJson, isRecord } from "@insightgraph/core";

export const VERIFIER_SYSTEM_PROMPT = `You are a verification agent for InsightGraph. Your job is to verify an analysis by checking:

1. Every key finding has supporting evidence from the retrieved data.
2. No contradictions exist between findings.
3. The confidence score is appropriate given the evidence.
4. Citations are accurate.

Respond in JSON:
{
  "verified": true,
  "issues": [],
  "adjusted_confidence": 0.85,
  "evidence_coverage": 0.9,
  "notes": "optional notes"
}

If issues are found:
{
  "verified": false,
  "issues": ["Finding X has no supporting evidence", "Contradiction between..."],
  "adjusted_confidence": 0.4,
  "evidence_coverage": 0.5,
  "notes": "..."
}`;

export interface VerificationResult {
  verified: boolean;
  issues: string[];
  adjustedConfidence: number;
  evidenceCoverage: number;
  notes: string;
}

/**
 * Checks conclusions for evidence support, contradictions, and confidence.
 */
export class Verifier {
  private client: LLMClient;
  private model: string;

  constructor(model = "gpt-4o-mini", apiKey = "", baseUrl = "") {
    this.client = createLLMClient(apiKey, baseUrl || undefined);
    this.model = model;
  }

  /**
   * Verify an analysis against the retrieved evidence.
   */
  async verify(
    analysis: Record<string, unknown>,
    retrievalResults: Array<Record<string, unknown>>,
  ): Promise<VerificationResult> {
    const context = JSON.stringify(
      { analysis, evidence: retrievalResults },
      null,
      2,
    );

    try {
      const raw = await chatJSON(
        this.client,
        this.model,
        [
          { role: "system", content: VERIFIER_SYSTEM_PROMPT },
          { role: "user", content: `Verify this analysis:\n${context}` },
        ],
        0.0,
      );
      const parsed = safeParseLlmJson<Record<string, unknown>>(raw, {
        context: "verifier",
        validate: isRecord,
      });
      if (!parsed) {
        return {
          verified: false,
          issues: ["Verifier response was not valid JSON"],
          adjustedConfidence: 0.0,
          evidenceCoverage: 0.0,
          notes: "Verification unavailable",
        };
      }
      return {
        verified: (parsed.verified as boolean) ?? false,
        issues: (parsed.issues as string[]) ?? [],
        adjustedConfidence: (parsed.adjusted_confidence as number) ?? 0.0,
        evidenceCoverage: (parsed.evidence_coverage as number) ?? 0.0,
        notes: (parsed.notes as string) ?? "",
      };
    } catch (err) {
      console.warn(`[verifier] chat failed: ${(err as Error).message}`);
      return {
        verified: false,
        issues: ["Verification failed due to an error"],
        adjustedConfidence: 0.0,
        evidenceCoverage: 0.0,
        notes: "Verification unavailable",
      };
    }
  }
}
